orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -1115,7 +1115,107 @@ void RetrieveSendRecvOperators(
   }
 }
 
-TEST(GradientGraphBuilderTest, PipelineOnlinePartition) {
+TEST(GradientGraphBuilderTest, PipelineOnlinePartition_bert_tiny) {
+  const auto model_path = ORT_TSTR("testdata/bert-tiny.onnx");
+
+  const size_t total_partition_count = 3;
+  TrainingSession::TrainingConfiguration::PipelineConfiguration pipe{};
+  pipe.do_partition = true;
+
+  // evenly model in 3 partitions
+  TrainingSession::TrainingConfiguration::CutInfo cut0 = {
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("186"),
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("71", {"273"})};
+
+  TrainingSession::TrainingConfiguration::CutInfo cut1 = {
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("308"),
+      onnxruntime::training::TrainingSession::TrainingConfiguration::CutEdge("71", {"395"})};
+
+  pipe.cut_list.emplace_back(cut0);
+  pipe.cut_list.emplace_back(cut1);
+
+  TrainingSession::TrainingConfiguration::MixedPrecisionConfiguration mixed_precision_config{};
+  mixed_precision_config.use_fp16_initializers = true;
+
+  // 2 test variations - full precision and mixed precision
+  const std::vector<bool> test_with_fp32{true, false};
+  for (auto is_fp32 : test_with_fp32) {
+    // graph is partitioned into 3 parts.
+    for (int i = 0; i < static_cast<int>(total_partition_count); ++i) {
```

Comment:
Think the static casts here can be avoided by "for (auto i=0u; ...)" or "for (size_t i=0u;...".

Reply:
`i` is later used to set world_rank and world_size, which are of type `int`. So it is either we declare `i` of type `int` and cast here or set `i` as `size_t` and cast when we assign the rank and size. I'm OK either way but the cast probably cannot be avoided... 

---
In reply to: [452242608](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r452242608) [](ancestors = 452242608)

Reply:
OK -- this sounds like something to revisit (separately from this PR).  I suspect we will want to align some of the different types used more generally.  Right now, across ORT, we often have a mix of int, unsigned, size_t, and ptrdiff_t which leads to "cast heavy" code.

Reply:
Agree. We can have a PR designated to unify some types and reduce the cast usage. 

---
In reply to: [453715077](https://github.com/microsoft/onnxruntime/pull/4321#discussion_r453715077) [](ancestors = 453715077)